sys/net/gnrc/link_layer/lwmac/rx_state_machine.c

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright (C) 2015 Daniel Krebs
+ *               2016 INRIA
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     net_lwmac
+ * @{
+ *
+ * @file
+ * @brief       Implementation of RX state machine of LWMAC protocol
+ *
+ * @author      Daniel Krebs <github@daniel-krebs.net>
+ * @author      Shuguo Zhuo  <shuguo.zhuo@inria.fr>
+ * @}
+ */
+
+#include <net/gnrc.h>
+#include <net/gnrc/lwmac/lwmac.h>
+#include <net/gnrc/mac/internal.h>
+
+#include "include/rx_state_machine.h"
+#include <net/gnrc/lwmac/timeout.h>
+#include "include/lwmac_internal.h"
+
+#define ENABLE_DEBUG    (1)
+#include "debug.h"
+
+#define LOG_LEVEL LOG_WARNING
+#include "log.h"
+
+#undef LOG_ERROR
+#undef LOG_WARNING
+#undef LOG_INFO
+#undef LOG_DEBUG
+
+#define LOG_ERROR(...) LOG(LOG_ERROR, "ERROR: [lwmac-rx] " __VA_ARGS__)
+#define LOG_WARNING(...) LOG(LOG_WARNING, "WARNING: [lwmac-rx] " __VA_ARGS__)
+#define LOG_INFO(...) LOG(LOG_INFO, "[lwmac-rx] " __VA_ARGS__)
+#define LOG_DEBUG(...) LOG(LOG_DEBUG, "[lwmac-rx] " __VA_ARGS__)
+
+/* Break out of switch and mark the need for rescheduling */
+#define GOTO_RX_STATE(rx_state, do_resched) gnrc_netdev2->rx.state = rx_state; \
```

Comment:
I acknowledge that hiding the `break` is bad, especially for the case where you might accidentally use it inside a loop.

> I'd rather have this inline, so I don't have to look it up every time to know whats happening

Well, this is actually why you abstract stuff away into functions (or defines for that matter), so you need *not* know whats happening behind the scenes. In this case I'd see this as an implementation detail of the state machine implementation, so I don't like `SET_RX_STATE_AND_BREAK` too much.

Maybe we have a valid case for using a `goto` here, like so:

```c
#define GOTO_RX_STATE(rx_state, do_resched) \
    gnrc_netdev2->rx.state = rx_state; \
    if(do_resched) goto rx_state_machine_begin; \
    else goto rx_state_machine_end

rx_state_machine_begin:
switch(rx.state)
{
// [...]
}
rx_state_machine_end:
```

What do you think? I don't like using `goto` too much either, but I've seen such an implementation sometimes. Probably the labels and `switch` statement should then also be wrapped inside macros.